cdap-explore/src/main/java/co/cask/cdap/explore/security/TokenSecureStoreUpdater.java

```diff
@@ -0,0 +1,161 @@
+      }
+
+      if (secureExplore) {
+        HiveTokenUtils.obtainToken(refreshedCredentials);
+        YarnTokenUtils.obtainToken(hConf, refreshedCredentials);
+        JobHistoryServerTokenUtils.obtainToken(hConf, refreshedCredentials);
+      }
+
+      YarnUtils.addDelegationTokens(hConf, locationFactory, refreshedCredentials);
+      credentials = refreshedCredentials;
+    } catch (IOException ioe) {
+      throw Throwables.propagate(ioe);
+    }
+  }
+
+  /**
+   * Since Hive classes are not in MasterServiceMain's classpath, create an instance of HiveConf using reflection.
+   * Call this method only if explore is enabled.
+   */
+  private Configuration getHiveConf() {
+    ClassLoader hiveClassloader = ExploreServiceUtils.getExploreClassLoader();
+    ClassLoader contextClassloader = Thread.currentThread().getContextClassLoader();
+    Thread.currentThread().setContextClassLoader(hiveClassloader);
+
+    try {
+      Class<?> clz = hiveClassloader.loadClass("org.apache.hadoop.hive.conf.HiveConf");
+      return (Configuration) clz.newInstance();
+    } catch (Exception e) {
+      LOG.error("Could not create an instance of HiveConf. Using default values.", e);
+      return null;
+    } finally {
+      Thread.currentThread().setContextClassLoader(contextClassloader);
+    }
+  }
+
+  private long calculateUpdateInterval() {
+    List<Long> renewalTimes = Lists.newArrayList();
+    // The value contains in hbase-default.xml, so it should always there. If it is really missing, default it to 1 day.
+    renewalTimes.add(hConf.getLong(Constants.HBase.AUTH_KEY_UPDATE_INTERVAL,
+                                   TimeUnit.MILLISECONDS.convert(1, TimeUnit.DAYS)));
+
+    if (secureExplore) {
+      // Renewal interval for YARN
+      renewalTimes.add(yarnConf.getLong(YarnConfiguration.DELEGATION_TOKEN_RENEW_INTERVAL_KEY,
+                                        YarnConfiguration.DELEGATION_TOKEN_RENEW_INTERVAL_DEFAULT));
+
+      // Renewal interval for Hive. Also see: https://issues.apache.org/jira/browse/HIVE-9214
+      Configuration hiveConf = getHiveConf();
+      if (hiveConf != null) {
+        renewalTimes.add(hiveConf.getLong(HadoopThriftAuthBridge20S.Server.DELEGATION_TOKEN_RENEW_INTERVAL_KEY,
+                                          HadoopThriftAuthBridge20S.Server.DELEGATION_TOKEN_RENEW_INTERVAL_DEFAULT));
+      } else {
+        renewalTimes.add(HadoopThriftAuthBridge20S.Server.DELEGATION_TOKEN_RENEW_INTERVAL_DEFAULT);
+      }
+
+      // Renewal interval for JHS
+      renewalTimes.add(hConf.getLong(MRConfig.DELEGATION_TOKEN_RENEW_INTERVAL_KEY,
+                                     MRConfig.DELEGATION_TOKEN_RENEW_INTERVAL_DEFAULT));
+    }
+
+    // Set the update interval to the shortest update interval of all required renewals.
+    Long minimumInterval = Collections.max(renewalTimes);
+    LOG.info("Setting token renewal time to: {}", minimumInterval);
+    return minimumInterval;
```

Comment:
It might be a good idea renew tokens a few minutes before they expire. See https://issues.apache.org/jira/browse/TWILL-101 for more information.